### PR TITLE
CLUSTERMON-2205: Added config support for Host-collectors 

### PR DESCRIPTION
**Pull request by Sprakhar97**
Refactored the cluster-controller code to improve code readability and maintainability
Broken down function of cluster-controller into cluster-collector and host-collector controllers
Combined the host-collector and cluster-collector into a common crd
Added Host-collector config field in the cluster-collectors cr to store configs for infra-agent and collectors in host-collector daemonset 
Made changes to supply configs to infra-agent and collectors present in host-collector daemonset through configmaps

### CODE REVIEW

File: pkg/controller/clustercollector/clusterCollector.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+			return reQueue, errUpdate
+		}
+
+		errRestart := c.RestartCollector()
+		if errRestart != nil {
+			reqLogger.Error(errRestart, "Failed to restart cluster collector", "clusterCollector.Namespace", clusterCollector.Namespace, "Deployment.Name", clusterCollector.Name)
+			return reQueue, errRestart
+		}
+	} else if updateDeployment {
+		fmt.Println("Breaking changes detected. Updating the the cluster collector deployment...")
```
Comment (subhashiyer07) [inline pkg/controller/clustercollector/clusterCollector.go:87]: Fix log - "No Breaking changes"

File: pkg/controller/clustercollector/collector_controller.go [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+version: %s
+clusterName: %s
+nsToMonitor: %s
+nsToExclude: %s
+clusterMonitoringEnabled: %t
+log-level: %s
+path: %s
+enabled: %t
+exporter-address: %s
+exporter-port: %d`, CLUSTER_COLLECTOR, TYPE_COLLECTOR, strings.Split(clusterCollector.Spec.Image, ":")[1], clusterCollector.Spec.ClusterName, clusterCollector.Spec.NsToMonitorRegex,
```
Comment (subhashiyer07) [inline pkg/controller/clustercollector/collector_controller.go:241]: version = strings.Split(clusterCollector.Spec.Image, ":")[1] may cause issues.
Like what if no version is specified in image and it defaults to latest?

Comment (subhashiyer07) [reply pkg/controller/clustercollector/collector_controller.go:241]: Same for other images

File: pkg/controller/clustercollector/utils.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+		if strings.Contains(protocol, "s") {
+			ssl = "true"
+		}
+		return nil, controllerDns, uint16(controllerPort), ssl
+	} else {
+		return fmt.Errorf("Controller Url is invalid. Use this format: protocol://dns:port"), "", 0, ""
+	}
+}
+
+func hasBreakingChanges(clusterCollector *appdynamicsv1alpha1.Clustercollector, existingDeployment *appsv1.Deployment) (bool, bool) {
```
Comment (subhashiyer07) [inline pkg/controller/clustercollector/utils.go:158]: nit: `hasBreakingChanges` on clusterCollector is defined here in utils while for hostCollector in hostCollector.go, with "different" implementations
We can try to make it consistent

File: pkg/controller/clustercollector/collector_controller.go (hunk 2 above)
Comment (Sprakhar97) [reply pkg/controller/clustercollector/collector_controller.go:241]: There is no other way for operator to know the version number, we can may be set up parameters such as version which are not available to operator through env variable directly in the image if IA consumes through env variable

**Review by subhashiyer07 — APPROVED** [review]

File: pkg/controller/clustercollector/collector_controller.go (hunk 2 above)
Comment (subhashiyer07) [reply pkg/controller/clustercollector/collector_controller.go]: Can we also initiate discussion with IA, if they can refer version from environment variable if available?
We can set the version as env in the image.
This way we can know the exact version and not use vague ones like `latest`, when version is not present in image tag

**Comment by harekal7** [comment]
We might have to remove the host-collector CRD from yaml?

**Review by harekal7 — APPROVED** [review]

**Pull request merged**